src/main/java/fi/aalto/cs/apluscourses/ui/exercise/ExercisesView.java

```diff
@@ -1,26 +1,52 @@
 package fi.aalto.cs.apluscourses.ui.exercise;
 
+import static fi.aalto.cs.apluscourses.utils.PluginResourceBundle.getText;
+
 import com.intellij.openapi.application.ApplicationManager;
 import com.intellij.openapi.application.ModalityState;
 import fi.aalto.cs.apluscourses.intellij.actions.ActionUtil;
 import fi.aalto.cs.apluscourses.intellij.actions.OpenSubmissionAction;
+import fi.aalto.cs.apluscourses.presentation.MainViewModel;
 import fi.aalto.cs.apluscourses.presentation.exercise.ExercisesTreeViewModel;
 import fi.aalto.cs.apluscourses.ui.GuiObject;
 import fi.aalto.cs.apluscourses.ui.base.TreeView;
+
+import java.awt.CardLayout;
+
+import javax.swing.JLabel;
 import javax.swing.JPanel;
+import javax.swing.JScrollPane;
+import javax.swing.SwingConstants;
+
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
+
 public class ExercisesView {
   private TreeView exerciseGroupsTree;
+  private JLabel emptyText;
   private JPanel basePanel;
-
+  private JScrollPane pane;
   @GuiObject
   public JPanel toolbarContainer;
+  private JPanel cardPanel;
+  private CardLayout cl;
+  private MainViewModel mainViewModel;
+  private boolean hasCourse = false;
```

Comment:
If you follow my suggestions, you shouldn't need this private field. In any case, I think passing information via mutable private fields (heap) should be avoided, if it can be passed in arguments of method calls (stack).